### PR TITLE
Use fallback libraries for archs without optimized logic (v2)

### LINKED ISSUE #1757: Tensile won't produce backend libraries for archs without optimized logic files when using --separate-architectures
# Issue
Tensile won't produce backend libraries for archs without optimized logic files when using --separate-architectures.

# Description
According with https://github.com/ROCmSoftwarePlatform/Tensile/issues/1165#issuecomment-1094556880 "gfx1010 has been enabled by default in rocBLAS builds since ROCm 4.3.0." however since rocBLAS does not have optimized logic files for navi10 no library is produced for gfx1010.

```shell
$ drun --rm rocm/dev-ubuntu-22.04:5.6-complete
root@ftl:/# ls -1 /opt/rocm/lib/rocblas/library/TensileLibrary_lazy_gfx*
/opt/rocm/lib/rocblas/library/TensileLibrary_lazy_gfx1030.dat
/opt/rocm/lib/rocblas/library/TensileLibrary_lazy_gfx1100.dat
/opt/rocm/lib/rocblas/library/TensileLibrary_lazy_gfx1101.dat
/opt/rocm/lib/rocblas/library/TensileLibrary_lazy_gfx1102.dat
/opt/rocm/lib/rocblas/library/TensileLibrary_lazy_gfx803.dat
/opt/rocm/lib/rocblas/library/TensileLibrary_lazy_gfx900.dat
/opt/rocm/lib/rocblas/library/TensileLibrary_lazy_gfx906.dat
/opt/rocm/lib/rocblas/library/TensileLibrary_lazy_gfx908.dat
/opt/rocm/lib/rocblas/library/TensileLibrary_lazy_gfx90a.dat
```
# Expected
Tensile should produce libraries for all requested architectures, using the fallback logic files for archs missing optimized logic files.

# Workaround
Building rocBLAS with `--merge-architectures --no-lazy-library-loading` seems to avoid the issue.

# Patch
https://github.com/ulyssesrr/docker-rocm-xtra/blob/3be41a9d79ff4f4324f3f34383b2282529c0c4b7/rocm-xtra-builder-rocblas/patches/Tensile-fix-fallback-arch-build.patch

### PR DESCRIPTION
**Pull request by GZGavinZhao**
Fixes #1757. Reintroducing #1862.

Enables architectures that don't have optimized logic files to also produce libraries when `--separate-architectures` or `--lazy-library-loading` is turned on. Previously, one must disable both of these two flags in order for rocBLAS to run on architectures like `gfx1010`.

Previously, there was a bug in Tensile solution indexing that caused #1862 to be reverted. Now, it seems like this issue has been fixed in #1888.

Test plan:
```
cmake -GNinja -B build -S . \
    -DCMAKE_C_COMPILER=hipcc \
    -DCMAKE_CXX_COMPILER=hipcc \
    -DBUILD_CLIENTS_TESTS=ON \
    -DBUILD_CLIENTS_BENCHMARKS=OFF \
    -DBUILD_CLIENTS_SAMPLES=OFF \
    -DBUILD_TESTING=ON \
    -DBUILD_WITH_TENSILE=ON \
    -DTensile_PRINT_DEBUG=ON \
    -DTensile_LIBRARY_FORMAT=msgpack \
    -DTensile_CPU_THREADS=14 \
    -DTensile_LAZY_LIBRARY_LOADING=ON \
    -DAMDGPU_TARGETS="..."
```
With `AMDGPU_TARGETS` being one of the following
- `AMDGPU_TARGETS=gfx1010`
- `AMDGPU_TARGETS=gfx1030;gfx1010`

In all cases,
`$ROCM_PATH/lib/rocblas/library/TensileLibrary_lazy_gfx1010.dat` is produced
and all other `*.dat` files remain unchanged.

In the second case, `./build/clients/staging/rocblas-test --gtest_filter='*gemm_ex_get_solutions*'` that previously failed now passes. I cannot run the full test suite due to limited memory on my GPU (I often get hipOutOfMemory when running stress tests). If this PR doesn't cause extra failures on AMD's CI or if someone can run the full test suite to ensure no additional failures are introduced, then I believe this PR should be good to go. Hopefully this PR can make it in before ROCm 6.1.

### CODE REVIEW

**Comment by hiepxanh** [comment]
Yes, you are my hope ❤️❤️❤️❤️❤️ @GZGavinZhao

**Comment by GZGavinZhao** [comment]
`TensileLibrary_gfx1010.co` is not produced while `TensileLibrary_gfx1030.co` is. Is this expected? I tried running tests emulating my `gfx1032` as `gfx1010` and it passed, so I think this is fine?

**Review by nakajee — APPROVED** [review]
Trial rocblas test all passed.

**Comment by wangxing7714436** [comment]
Anyone you guys could upload gfx1010.dat before this pr merge into rocm,please? couldn't wait for running ollama with my 5700xt. Many Thanks.

**Comment by GZGavinZhao** [comment]
@wangxing7714436 I can't guarantee it will work, but I can give it a try. What ROCm version do you have?

**Comment by wangxing7714436** [comment]
> @wangxing7714436 I can't guarantee it will work, but I can give it a try. What ROCm version do you have?

I installed 5.7 on windows, no tensile file on gfx1010. Many thanks for your reply.

**Comment by GZGavinZhao** [comment]
@wangxing7714436 Uh this is a little tricky. I'm not familiar with how rocBLAS works on Windows. If you **can't** find files like `TensileLibrary_Type_4xi8I_HPA_Contraction_l_Alik_Bjlk_Cijk_Dijk_fallback_gfx1010.hsaco`, then I'm almost certain this won't work for you. If you can find these files, then in the same directory where you found these files, extract then put the attached `.dat` file there and see if it will work. If this doesn't work, then I think you would unfortunately have to wait until the next ROCm Windows SDK release.
 
[TensileLibrary_lazy_gfx1010.dat.zip](https://github.com/user-attachments/files/16163193/TensileLibrary_lazy_gfx1010.dat.zip)

**Pull request merged**
